### ISSUE TITLE
Allow Map types without 'map' attribute

### ISSUE THREAD

**Issue by q-ode** [dynamoosejs/dynamoose#541]
### Summary:
I've looked through pretty much all the issues on how the library handles `Map` types and I've found what I believe could be an additional benefit here.

As pointed out in a comment under #230 , using `Map` type in combination with setting `useDocumentTypes: true` you'd be able to define a schema for your map using the `map` keyword and the library would save it as a map in DynamoDB properly.

This won't work for maps without predefined fields. Using `saveUnknown: true` doesn't work either as it the field ends up being a string or empty object depending on the combination with `useDocumentTypes` in DynamoDB as mentioned earlier in #146.

#### Schema
```
const sampleSchema = new Schema({
  id: String,
  blackbox: Map
}, {
  timestamps: true,
  useDocumentTypes: true,
});
```

#### General
I did some digging around the codebase and was able to make a likely fix here: https://github.com/dynamoosejs/dynamoose/blob/a55b11f03a35bb5659147c60bd281a0c88591cbc/lib/Attribute.js#L474

If a change of this nature was made then we would be able to save maps without pre-defined attributes and it would be saved as an actual Map on DynamoDB. Would of course need to take care of nested objects/maps:
```
let dynamoMapObj = {};
    if (Object.keys(this.attributes).length) {
      for(const name in this.attributes) {
        const attr = this.attributes[name];
        await attr.setDefault(model);
        const dynamoAttr = await attr.toDynamo(val[name], undefined, model);
        if(dynamoAttr) {
          dynamoMapObj[attr.name] = dynamoAttr;
        }
      }
    } else {
      for (const key in val) {
        dynamoMapObj[key] = {
          S: val[key]
        };
      }
    }
    dynamoObj.M = dynamoMapObj;
```

### Dynamoose Plugins:

- [ ] Yes, I believe that one or multiple 3rd party Dynamoose plugins are effecting this issue
- [ ] No, I believe this issue is independent of any 3rd party Dynamoose plugins I'm using
- [ ] Unknown, I'm unsure if Dynamoose plugins are effecting this issue
- [ ] I am not using any Dynamoose plugins

### Type (select 1):
- [ ] Bug report
- [ ] Feature suggestion
- [ ] Question
- [ ] Other suggestion
- [ ] Something not listed here

### Other:
- [ ] I have read through the Dynamoose documentation before posting this issue
- [ ] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [ ] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [ ] I have tested the code provided and am confident it doesn't work as intended
- [ ] I have ensured that all of my plugins that I'm using with Dynamoose are listed above
- [ ] I have filled out all fields above
- [ ] I am running the latest version of Dynamoose

**Comment by q-ode** [comment]
> Using saveUnknown: true doesn't work either as it the field ends up being a string or empty object depending on the combination with useDocumentTypes in DynamoDB as mentioned earlier in #146.

I had mentioned in the earlier comment that save unknown doesn't finally translate the document into a Map as expected DynamoDB. You end up with an empty object thereby loosing data.

This is a code sample:
```
const sampleSchema = new Schema({
  id: String,
  blackbox: Map,
}, {
  timestamps: true,
  useDocumentTypes: true,
  saveUnknown: true
});

var sampleModel = dynamoose.model('dev-table-sample', sampleSchema);

new sampleModel({id: Date.now(), blackbox: {name: 'John Doe'}}).save();
```

This what is stored in DynamoDB:
<img width="351" alt="screen shot 2019-01-18 at 10 37 30 am" src="https://user-images.githubusercontent.com/929803/51378273-363cf480-1b0d-11e9-9392-504f54f7cbe0.png">

**Comment by fishcharlie** [comment]
@q-ode Yeah. The solution might be to throw an error when you don’t pass a map property in. If you remove that property from the schema with saveUnknown turned on, it should work.
